### PR TITLE
fix: align bloom filter size/hash calculation with Dash Core

### PR DESCRIPTION
**Pull request by xdustinface**
Fix `BloomFilter` size calculation to compute filter size in bytes first (integer division), then convert to bits (bytes*8), matching how Dash Core implementation sizes the filter.

The previous code used `ceil` on bits directly, leading to wrong filter sizes and hash moduli that caused bloom filter mismatches.

## Summary by CodeRabbit

* **Bug Fixes**
  * Bloom filter sizing and hashing behavior updated to match Dash Core, resolving edge cases (including very small/1-byte filters) so filters interoperate correctly and reliably.

* **Tests**
  * Added extensive tests that validate Bloom filter serialization and operations against Dash Core reference vectors, including scenarios with key insertions to ensure compatibility and correctness.

### CODE REVIEW

File: dash/src/bloom/filter.rs [hunk 1]
```diff
@@ -294,4 +305,113 @@ mod tests {
...
+            encoded, expected,
+            "Serialized bloom filter must match Dash Core bloom_create_insert_key test vector"
+        );
+    }
+
+    /// Verify with tweak = 2147483649 from Dash Core's bloom_create_insert_serialize_with_tweak:
+    /// https://github.com/dashpay/dash/blob/b66b56c3019fe7ab0c9f35dd9894c0ded4d9d420/src/test/bloom_tests.cpp#L51-L74
+    #[test]
+    fn test_bloom_filter_dash_core_compatibility_with_tweak() {
+        let mut filter = BloomFilter::new(3, 0.01, 2147483649, BloomFlags::All).unwrap();
```
Comment (ZocoLini) [inline dash/src/bloom/filter.rs:387]: I dont see difference beetwen this test and `test_bloom_filter_dash_core_compatibility` other than the filter constructor, maybe a for loop over a collection containing both bloom filters?? Maybe extract the logic into a function encapsulated in the test function and call it with both filters??

File: dash/src/bloom/filter.rs [hunk 2]
```diff
@@ -50,23 +50,26 @@ impl BloomFilter {
...
-        let filter_bytes = filter_size.div_ceil(8);
         if filter_bytes > MAX_BLOOM_FILTER_SIZE {
             return Err(BloomError::FilterTooLarge(filter_bytes));
         }
 
-        let filter = bitvec![u8, Lsb0; 0; filter_size];
+        let aligned_bits = filter_bytes * 8;
+
+        let n_hash_funcs = (aligned_bits as f64 / elements as f64 * ln2) as u32;
+        let n_hash_funcs = n_hash_funcs.clamp(1, MAX_HASH_FUNCS);
```
Comment (ZocoLini) [inline dash/src/bloom/filter.rs:70]: using clamp is kinda unnecessary, u32 cannot be less than 1 and the C++ code is not checking it either

File: dash/src/bloom/filter.rs [hunk 3]
```diff
@@ -50,23 +50,26 @@ impl BloomFilter {
...
+        // https://github.com/dashpay/dash/blob/1c9809f1c6b917a88e466e2e059e0594afb04a59/src/common/bloom.cpp#L29-L45
         let ln2 = std::f64::consts::LN_2;
         let ln2_squared = ln2 * ln2;
 
-        let filter_size =
-            (-(elements as f64) * false_positive_rate.ln() / ln2_squared).ceil() as usize;
-        let filter_size = filter_size.clamp(1, MAX_BLOOM_FILTER_SIZE * 8);
+        let filter_bits =
+            (-1.0 / ln2_squared * elements as f64 * false_positive_rate.ln()) as usize;
+        let filter_bits = filter_bits.clamp(8, MAX_BLOOM_FILTER_SIZE * 8);
```
Comment (ZocoLini) [inline dash/src/bloom/filter.rs]: I dont see the C++ code checking if the value is less than 8, and I dont know enough about BloomFilters implementation right now, so I have to trust you

File: dash/src/bloom/filter.rs [hunk 4]
```diff
@@ -294,4 +305,113 @@ mod tests {
...
+    /// Verify bloom filter with pubkey and pubkey hash insertion matches Dash Core's
+    /// bloom_create_insert_key test:
+    /// https://github.com/dashpay/dash/blob/b66b56c3019fe7ab0c9f35dd9894c0ded4d9d420/src/test/bloom_tests.cpp#L76-L95
+    ///
+    /// Filter: 2 elements, 0.001 fp rate, tweak 0, BLOOM_UPDATE_ALL.
+    /// Data inserted: uncompressed pubkey bytes and Hash160(pubkey) bytes.
+    /// Expected serialized bytes: "038fc16b080000000000000001"
+    #[test]
+    fn test_bloom_create_insert_key() {
+        let secp = secp256k1::Secp256k1::new();
```
Comment (ZocoLini) [inline dash/src/bloom/filter.rs:350]: This unit test is testing the same paths (in BloomFilter) the `test_bloom_filter_dash_core_compatibility` function is testing, right??

**Review by ZocoLini — CHANGES_REQUESTED** [review]

File: dash/src/bloom/filter.rs (hunk 2 above)
Comment (xdustinface) [reply dash/src/bloom/filter.rs:70]: > u32 cannot be less than 1

What about 0? Apparently unlikely to happen but maybe just an edge case issue in dash core then.

File: dash/src/bloom/filter.rs (hunk 1 above)
Comment (xdustinface) [reply dash/src/bloom/filter.rs:387]: Its just mirroring the c++ test cases.

File: dash/src/bloom/filter.rs (hunk 4 above)
Comment (xdustinface) [reply dash/src/bloom/filter.rs:350]: Its just mirroring the c++ test cases.

File: dash/src/bloom/filter.rs (hunk 3 above)
Comment (xdustinface) [reply dash/src/bloom/filter.rs]: Hm right, going to drop it.

File: dash/src/bloom/filter.rs (hunk 2 above)
Comment (ZocoLini) [reply dash/src/bloom/filter.rs:70]: omg, for a moment i was considering the mathematic definition of positive in my head xdddd, mb

File: dash/src/bloom/filter.rs (hunk 1 above)
Comment (ZocoLini) [reply dash/src/bloom/filter.rs:387]: as you wish

**Review by ZocoLini — APPROVED** [review]
